Null Pointer Exception in ConfigInvoker.setMethodInvocationFailure

### TestNG Version
> Note: only the latest version is supported

7.1.0

### Expected behavior
When running tests, ConfigInvoker.setMethodInvocationFailure should not be throwing Null Pointer Exception

### Actual behavior
A Null Pointer Exception is thrown by TestNG:

Exception java.lang.NullPointerException
Message: (none)

Stacktrace:

```
at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1645)
at org.testng.internal.ConfigInvoker.setMethodInvocationFailure(ConfigInvoker.java:488)
at org.testng.internal.ConfigInvoker.recordConfigurationInvocationFailed(ConfigInvoker.java:529)
at org.testng.internal.ConfigInvoker.handleConfigurationFailure(ConfigInvoker.java:441)
at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:308)
at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
```

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
> Please, share the test case (as small as possible) which shows the issue

I've been unable to narrow down the exact cause thus far. I believe it might happen when there is an exception in a setup method or parent setup method (while running the tests in parallel).  However I am still looking into it.

Either way, it seems like we should add a null check in java.util.concurrent.ConcurrentHashMap.computeIfAbsent to prevent this issue from happening.  I believe this NPE is covering up the actual exception being thrown by the tests.

@krmahadevan  - I updated the groups_at_test_level.xml file to be:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
<suite thread-count="256" configfailurepolicy="continue" verbose="2" name="2232_suite_groups_at_test_level"
  parallel="classes">
  <test thread-count="256" verbose="2" name="2232_test" parallel="classes">
    <groups>
      <run>
        <include name="Group2"/> <exclude name="Broken"/>
      </run>
    </groups>
    <!--classes>
      <class name="com.rationaleemotions.SampleTest"/>
      <class name="com.rationaleemotions.SampleTest2"/>
    </classes-->

    <packages>
      <package name="com.rationaleemotions.*" />
    </packages>
  </test> <!-- 2232_test -->
</suite> <!-- 2232_suite -->
```

and ran using this command:
```
mvn clean install test -Dsurefire.suiteXmlFiles='src/test/resources/groups_at_test_level.xml'
```

This is the output I got (might need to run it a few times to make sure the exception is called):
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
...
... TestNG 7.1.0 by Cédric Beust (cedric@beust.com)
...

Setup1 variable1=0
Setup2 variable2=0
Setup2 variable2=1
Teardown2 variable2=2
Setup1 variable1=1
Teardown1 variable1=2
Test1 variable1=3
Setup2 variable2=3
Teardown1 variable1=4
Test2 variable1=5
Teardown1 variable1=6
FAILED CONFIGURATION: setUp
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:17)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

FAILED CONFIGURATION: tearDown
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1645)
	at org.testng.internal.ConfigInvoker.setMethodInvocationFailure(ConfigInvoker.java:488)
	at org.testng.internal.ConfigInvoker.recordConfigurationInvocationFailed(ConfigInvoker.java:529)
	at org.testng.internal.ConfigInvoker.handleConfigurationSkip(ConfigInvoker.java:399)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:268)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

PASSED: test1("Test1")
PASSED: test1("Test2")
FAILED: test2
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1645)
	at org.testng.internal.ConfigInvoker.setMethodInvocationFailure(ConfigInvoker.java:488)
	at org.testng.internal.ConfigInvoker.recordConfigurationInvocationFailed(ConfigInvoker.java:529)
	at org.testng.internal.ConfigInvoker.handleConfigurationFailure(ConfigInvoker.java:441)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:308)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


===============================================
    2232_test
    Tests run: 3, Failures: 1, Skips: 0
    Configuration Failures: 2, Skips: 0
===============================================

[ERROR] Tests run: 5, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 0.516 s <<< FAILURE! - in TestSuite
[ERROR] setUp(com.rationaleemotions.SampleTest2)  Time elapsed: 0.003 s  <<< FAILURE!
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:17)

[ERROR] tearDown(com.rationaleemotions.SampleTest2)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.NullPointerException

[ERROR] test2(com.rationaleemotions.SampleTest2)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.NullPointerException

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   SampleTest2.setUp:17  Exception!
[ERROR]   SampleTest2.tearDown » NullPointer
[ERROR]   SampleTest2.test2 » NullPointer
[INFO] 
[ERROR] Tests run: 5, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.518 s
[INFO] Finished at: 2020-02-20T13:37:21-08:00
[INFO] ------------------------------------------------------------------------

```

Please let me know if you need any further information.  Thanks!

@krmahadevan  - I am using this Java version (on Mac):
```
$ java -version
java version "1.8.0_162"
Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.162-b12, mixed mode)
```

I am able to reproduce the NPE in ConfigInvoker.setMethodInvocationFailure on both Mac and Linux.

I deleted everything in the `~/.m2/repository/org/testng` folder, then ran `mvn dependency:resolve`.  Then I ran the tests using this command on the command line (Mac):
`mvn clean install test -Dsurefire.suiteXmlFiles='src/test/resources/groups_at_test_level.xml'`

and got this output:

```
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< com.rationaleemotions:issue_2232 >------------------
[INFO] Building issue_2232 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ issue_2232 ---
[INFO] Deleting /issue_2232-2/target
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ issue_2232 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /issue_2232-2/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ issue_2232 ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ issue_2232 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ issue_2232 ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /issue_2232-2/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ issue_2232 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
...
... TestNG 7.1.0 by Cédric Beust (cedric@beust.com)
...

Setup1 variable1=0
Setup2 variable2=0
Setup2 variable2=1
Teardown2 variable2=2
Setup1 variable1=1
Teardown1 variable1=2
Test1 variable1=3
Setup2 variable2=3
Teardown1 variable1=4
Test2 variable1=5
Teardown1 variable1=6
FAILED CONFIGURATION: setUp
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:17)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

FAILED CONFIGURATION: tearDown
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1645)
	at org.testng.internal.ConfigInvoker.setMethodInvocationFailure(ConfigInvoker.java:488)
	at org.testng.internal.ConfigInvoker.recordConfigurationInvocationFailed(ConfigInvoker.java:529)
	at org.testng.internal.ConfigInvoker.handleConfigurationSkip(ConfigInvoker.java:399)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:268)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

PASSED: test1("Test1")
PASSED: test1("Test2")
FAILED: test2
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1645)
	at org.testng.internal.ConfigInvoker.setMethodInvocationFailure(ConfigInvoker.java:488)
	at org.testng.internal.ConfigInvoker.recordConfigurationInvocationFailed(ConfigInvoker.java:529)
	at org.testng.internal.ConfigInvoker.handleConfigurationFailure(ConfigInvoker.java:441)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:308)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


===============================================
    2232_test
    Tests run: 3, Failures: 1, Skips: 0
    Configuration Failures: 2, Skips: 0
===============================================

[ERROR] Tests run: 5, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 0.611 s <<< FAILURE! - in TestSuite
[ERROR] setUp(com.rationaleemotions.SampleTest2)  Time elapsed: 0.003 s  <<< FAILURE!
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:17)

[ERROR] tearDown(com.rationaleemotions.SampleTest2)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.NullPointerException

[ERROR] test2(com.rationaleemotions.SampleTest2)  Time elapsed: 0.004 s  <<< FAILURE!
java.lang.NullPointerException

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   SampleTest2.setUp:17  Exception!
[ERROR]   SampleTest2.tearDown » NullPointer
[ERROR]   SampleTest2.test2 » NullPointer
[INFO] 
[ERROR] Tests run: 5, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.017 s
[INFO] Finished at: 2020-02-21T11:08:06-08:00
[INFO] ------------------------------------------------------------------------
```

What command are you using to run the tests?  What is the output you are seeing?  You may have to run the tests 4 or 5 times to see the error.

Also, did you update the xml file mentioned above?

Thanks.

@bc-joshroe - 

On my MAC, I am running on 

```
11:51 $ java -version
java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
```

And on my Ubuntu VM, I am running on 

```
$:~/issue_2232$ java --version
openjdk 11.0.6 2020-01-14
OpenJDK Runtime Environment (build 11.0.6+10-post-Ubuntu-1ubuntu119.10.1)
OpenJDK 64-Bit Server VM (build 11.0.6+10-post-Ubuntu-1ubuntu119.10.1, mixed mode, sharing)
```
Can you try upgrading your JDK to latest version in JDK8 ? You seem to be on an older version (Just trying to negate any bugs in JDK that may be causing this)


> What command are you using to run the tests? 

`mvn clean test`

> What is the output you are seeing? You may have to run the tests 4 or 5 times to see the error.

I can simulate the test configuration throwing the `Exception` , but I cant simulate TestNG triggering an NPE (which is what you are seeing.

```
$ mvn clean test
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------< com.rationaleemotions:issue_2232 >------------------
[INFO] Building issue_2232 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ issue_2232 ---
[INFO] Deleting /home/krmahadevan/issue_2232/target
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ issue_2232 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/krmahadevan/issue_2232/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ issue_2232 ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ issue_2232 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ issue_2232 ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /home/krmahadevan/issue_2232/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ issue_2232 ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
...
... TestNG 7.0.1 by CÃ©dric Beust (cedric@beust.com)
...

Setup1 variable1=0
Setup2 variable2=0
Setup1 variable1=1
Teardown1 variable1=2
Setup2 variable2=1
Teardown2 variable2=2
Setup1 variable1=3
Test1 variable1=4
Teardown1 variable1=5
Setup2 variable2=3
Teardown1 variable1=6
Test2 variable1=7
Teardown2 variable2=4
Teardown1 variable1=8
Setup1 variable1=9
Teardown1 variable1=10
Setup2 variable2=5
Teardown2 variable2=6
Setup1 variable1=11
Teardown1 variable1=12
Teardown2 variable2=7
Setup2 variable2=8
FAILED CONFIGURATION: setUp
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:20)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

SKIPPED CONFIGURATION: @AfterMethod tearDown
SKIPPED CONFIGURATION: @AfterMethod tearDown
PASSED: test1("Test1")
PASSED: test1("Test2")
SKIPPED: test2
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:20)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

SKIPPED: test2
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:20)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.ConfigInvoker.invokeBeforeGroupsConfigurations(ConfigInvoker.java:142)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:518)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)


===============================================
    2232_test
    Tests run: 4, Failures: 0, Skips: 2
    Configuration Failures: 1, Skips: 2
===============================================

[ERROR] Tests run: 7, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 0.938 s <<< FAILURE! - in TestSuite
[ERROR] setUp(com.rationaleemotions.SampleTest2)  Time elapsed: 0.002 s  <<< FAILURE!
java.lang.Exception: Exception!
	at com.rationaleemotions.SampleTest2.setUp(SampleTest2.java:20)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   SampleTest2.setUp:20  Exception!
[INFO]
[ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 2
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.674 s
[INFO] Finished at: 2020-02-23T12:03:46+05:30
[INFO] ------------------------------------------------------------------------
```



 > Also, did you update the xml file mentioned above?

Yes I have added the `exclude` tag.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2232: Null Pointer Exception in ConfigInvoker.setMethodInvocationFailure (Krishnan Mahadevan)
 Fixed: GITHUB-2312: IAnnotationTransformer called multiple time Discrepancy between 6.x and 7.x (Krishnan Mahadevan)
 New:   GITHUB-2315: TextReporter console output does not nicely print native array data parameters (James Sassano)
 Fixed: GITHUB-2301: Add support for object-based reporter configurations (Scott Babcock)

--- a/src/main/java/org/testng/internal/ConfigInvoker.java
+++ b/src/main/java/org/testng/internal/ConfigInvoker.java
@@ -144,6 +144,7 @@ class ConfigInvoker extends BaseInvoker implements IConfigInvoker {
           .forSuite(arguments.getSuite())
           .usingParameters(arguments.getParameters())
           .usingInstance(arguments.getInstance())
+          .forTestMethod(arguments.getTestMethod())
           .build();
       invokeConfigurations(configMethodArguments);
     }
@@ -216,6 +217,7 @@ class ConfigInvoker extends BaseInvoker implements IConfigInvoker {
         .forSuite(arguments.getSuite())
         .usingParameters(arguments.getParameters())
         .usingInstance(arguments.getInstance())
+        .forTestMethod(arguments.getTestMethod())
         .build();
 
     invokeConfigurations(configMethodArguments);
@@ -500,6 +502,9 @@ class ConfigInvoker extends BaseInvoker implements IConfigInvoker {
   }
 
   private void setMethodInvocationFailure(ITestNGMethod method, Object instance) {
+    if (method == null) {
+      return;
+    }
     Set<Object> instances =
         m_methodInvocationResults.computeIfAbsent(method, k -> Sets.newHashSet());
     instances.add(TestNgMethodUtils.getMethodInvocationToken(method, instance));

--- a/src/test/java/test/groups/issue2232/IssueTest.java
+++ b/src/test/java/test/groups/issue2232/IssueTest.java
@@ -1,0 +1,92 @@
+package test.groups.issue2232;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import org.testng.Reporter;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlGroups;
+import org.testng.xml.XmlPackage;
+import org.testng.xml.XmlRun;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+import test.SimpleBaseTest;
+
+import java.util.Collections;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+public class IssueTest extends SimpleBaseTest {
+
+  private static final Random random = new Random();
+
+  @Test(description = "GITHUB-2232", invocationCount = 10)
+  //This test case doesn't vet out the fix completely because the bug by itself is very
+  //sporadic and is not easy to reproduce. That is why this test is being executed 10 times
+  // to ensure that the issue can be reproduced in one of the executions
+  public void ensureNoNPEThrownWhenRunningGroups() throws InterruptedException {
+    TestNG testng = create(constructSuite());
+    testng.run();
+    assertThat(testng.getStatus()).isEqualTo(0);
+    TimeUnit.MILLISECONDS.sleep(random.nextInt(5000));
+  }
+
+  private XmlSuite constructSuite() {
+    XmlSuite xmlsuite = createXmlSuite("2232_suite");
+    xmlsuite.setConfigFailurePolicy(XmlSuite.FailurePolicy.CONTINUE);
+    xmlsuite.setThreadCount(256);
+    xmlsuite.setVerbose(2);
+    xmlsuite.setParallel(XmlSuite.ParallelMode.CLASSES);
+    XmlTest xmltest = createXmlTest(xmlsuite, "2232_test");
+    XmlRun xmlrun = new XmlRun();
+    xmlrun.onInclude("Group2");
+    xmlrun.onExclude("Broken");
+    XmlGroups xmlgroup = new XmlGroups();
+    xmlgroup.setRun(xmlrun);
+    xmltest.setGroups(xmlgroup);
+    XmlPackage xmlpackage = new XmlPackage();
+    xmlpackage.setName(getClass().getPackage().getName() + ".samples.*");
+    xmltest.setPackages(Collections.singletonList(xmlpackage));
+    return xmlsuite;
+  }
+
+  @Test(invocationCount = 10, description = "GITHUB-2232")
+  //Ensuring that the bug doesn't surface even when tests are executed via the command line mode
+  public void commandlineTest() throws IOException, InterruptedException {
+    Path suitefile = Files.write(Files.createTempFile("testng", ".xml"),
+        constructSuite().toXml().getBytes(Charset.defaultCharset()));
+    List<String> args = Collections.singletonList(suitefile.toFile().getAbsolutePath());
+    int status = exec(Collections.emptyList(), args);
+    assertThat(status).isEqualTo(0);
+    TimeUnit.MILLISECONDS.sleep(random.nextInt(5000));
+  }
+
+  private int exec(List<String> jvmArgs, List<String> args)
+      throws IOException, InterruptedException {
+
+    String javaHome = System.getProperty("java.home");
+    String javaBin = javaHome + File.separator + "bin" + File.separator + "java";
+    String classpath = System.getProperty("java.class.path");
+    String className = TestNG.class.getName();
+    List<String> command = new ArrayList<>();
+    command.add(javaBin);
+    command.addAll(jvmArgs);
+    command.add("-cp");
+    command.add(classpath);
+    command.add(className);
+    command.addAll(args);
+    Reporter.log("Executing the command " + command, true);
+    ProcessBuilder builder = new ProcessBuilder(command);
+    Process process = builder.inheritIO().start();
+    process.waitFor();
+
+    return process.exitValue();
+  }
+}

--- a/src/test/java/test/groups/issue2232/samples/SampleTest.java
+++ b/src/test/java/test/groups/issue2232/samples/SampleTest.java
@@ -1,0 +1,35 @@
+package test.groups.issue2232.samples;
+
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+@Test(groups={"Group1", "Group2", "Group3"})
+public class SampleTest {
+
+    private int variable = 0;
+
+    @BeforeClass
+    public void setUp() {
+        variable += 1;
+    }
+
+    @AfterMethod
+    public void tearDown() {
+        variable += 1;
+    }
+
+    @DataProvider(name="testData1")
+    public Object[][] testData1() {
+        return new Object[][] {
+                {"Test1"},
+                {"Test2"},
+        };
+    }
+
+    @Test(dataProvider="testData1")
+    public void test1(String test) {
+        variable+=1;
+    }
+}

--- a/src/test/java/test/groups/issue2232/samples/SampleTest2.java
+++ b/src/test/java/test/groups/issue2232/samples/SampleTest2.java
@@ -1,0 +1,38 @@
+package test.groups.issue2232.samples;
+
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+@Test(groups={"Group1", "Group2", "Group3"})
+public class SampleTest2 {
+
+    private int variable = 0;
+
+    @BeforeClass
+    public void setUp() throws Exception {
+        variable+=1;
+        if (variable == 4) {
+            throw new Exception("Exception!");
+        }
+    }
+
+    @AfterMethod
+    public void tearDown() {
+        variable+=1;
+    }
+
+    @DataProvider(name="testData2")
+    public Object[][] testData2() {
+        return new Object[][] {
+                {"Test3"},
+                {"Test4"},
+        };
+    }
+
+    @Test(dataProvider="testData2")
+    public void test2(String test) {
+        variable+=1;
+    }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -156,6 +156,7 @@
       <class name="test.groups.issue1834.IssueTest"/>
       <class name="test.groups.issue182.IssueTest"/>
       <class name="test.groups.issue2152.IssueTest"/>
+      <class name="test.groups.issue2232.IssueTest"/>
       <class name="test.parameters.ShadowTest" />
       <class name="test.parameters.ParameterOverrideTest" />
       <class name="test.parameters.ParameterInjectAndOptionTest" />

